README.md
# DAWG-Python

[![Python tests](https://github.com/pymorphy2-fork/DAWG-Python/actions/workflows/python-tests.yml/badge.svg)](https://github.com/pymorphy2-fork/DAWG-Python/actions/workflows/python-tests.yml)
[![Coverage Status](https://coveralls.io/repos/github/pymorphy2-fork/DAWG-Python/badge.svg?branch=master)](https://coveralls.io/github/pymorphy2-fork/DAWG-Python?branch=master)

This pure-python package provides read-only access for files created by
[dawgdic](https://code.google.com/p/dawgdic/) C++ library and
[DAWG](https://github.com/kmike/DAWG) python package.

This package is not capable of creating DAWGs. It works with DAWGs built
by [dawgdic](https://code.google.com/p/dawgdic/) C++ library or
[DAWG](https://github.com/kmike/DAWG) Python extension module. The main
purpose of DAWG-Python is to provide access to DAWGs without
requiring compiled extensions. It is also quite fast under PyPy (see
benchmarks).

# Installation

pip install DAWG-Python

# Usage

The aim of DAWG-Python is to be API- and binary-compatible with
[DAWG](https://github.com/kmike/DAWG) when it is possible.

First, you have to create a dawg using
[DAWG](https://github.com/kmike/DAWG) module:

    import dawg
    d = dawg.DAWG(data)
    d.save('words.dawg')

And then this dawg can be loaded without requiring C extensions:

    import dawg_python
    d = dawg_python.DAWG().load('words.dawg')

Please consult [DAWG](https://github.com/kmike/DAWG) docs for detailed
usage. Some features (like constructor parameters or `save` method) are
intentionally unsupported.

# Benchmarks

Benchmark results (100k unicode words, integer values (lengths of the
words), PyPy 1.9, macbook air i5 1.8 Ghz):

    dict __getitem__ (hits):        11.090M ops/sec
    DAWG __getitem__ (hits):        not supported
    BytesDAWG __getitem__ (hits):   0.493M ops/sec
    RecordDAWG __getitem__ (hits):  0.376M ops/sec

    dict get() (hits):              10.127M ops/sec
    DAWG get() (hits):              not supported
    BytesDAWG get() (hits):         0.481M ops/sec
    RecordDAWG get() (hits):        0.402M ops/sec
    dict get() (misses):            14.885M ops/sec
    DAWG get() (misses):            not supported
    BytesDAWG get() (misses):       1.259M ops/sec
    RecordDAWG get() (misses):      1.337M ops/sec

    dict __contains__ (hits):           11.100M ops/sec
    DAWG __contains__ (hits):           1.317M ops/sec
    BytesDAWG __contains__ (hits):      1.107M ops/sec
    RecordDAWG __contains__ (hits):     1.095M ops/sec

    dict __contains__ (misses):         10.567M ops/sec
    DAWG __contains__ (misses):         1.902M ops/sec
    BytesDAWG __contains__ (misses):    1.873M ops/sec
    RecordDAWG __contains__ (misses):   1.862M ops/sec

    dict items():           44.401 ops/sec
    DAWG items():           not supported
    BytesDAWG items():      3.226 ops/sec
    RecordDAWG items():     2.987 ops/sec
    dict keys():            426.250 ops/sec
    DAWG keys():            not supported
    BytesDAWG keys():       6.050 ops/sec
    RecordDAWG keys():      6.363 ops/sec

    DAWG.prefixes (hits):    0.756M ops/sec
    DAWG.prefixes (mixed):   1.965M ops/sec
    DAWG.prefixes (misses):  1.773M ops/sec

    RecordDAWG.keys(prefix="xxx"), avg_len(res)==415:       1.429K ops/sec
    RecordDAWG.keys(prefix="xxxxx"), avg_len(res)==17:      36.994K ops/sec
    RecordDAWG.keys(prefix="xxxxxxxx"), avg_len(res)==3:    121.897K ops/sec
    RecordDAWG.keys(prefix="xxxxx..xx"), avg_len(res)==1.4: 265.015K ops/sec
    RecordDAWG.keys(prefix="xxx"), NON_EXISTING:            2450.898K ops/sec

Under CPython expect it to be about 50x slower. Memory consumption of
DAWG-Python should be the same as of
[DAWG](https://github.com/kmike/DAWG).

# Current limitations

-   This package is not capable of creating DAWGs;
-   all the limitations of [DAWG](https://github.com/kmike/DAWG) apply.

Contributions are welcome!

# Contributing

Development happens at GitHub: <https://github.com/pymorphy2-fork/DAWG-Python>
Issue tracker: <https://github.com/pymorphy2-fork/DAWG-Python/issues>

Feel free to submit ideas, bugs or pull requests.

## Running tests and benchmarks

Make sure [tox](http://tox.testrun.org) is installed and run

    $ tox

from the source checkout. Tests should pass under python 3.8, 3.9, 3.10, 3.11 and PyPy3 \>= 7.3.

In order to run benchmarks, type

    $ pypy3 -m bench.speed

This runs benchmarks under PyPy (they are about 50x slower under
CPython).

## Authors & Contributors

-   Mikhail Korobov \<<kmike84@gmail.com>\>

The algorithms are from [dawgdic](https://code.google.com/p/dawgdic/)
C++ library by Susumu Yata & contributors.

# License

This package is licensed under MIT License.


CHANGES.md
# Changes

## 0.7.2 (2015-04-18)

-   minor speedup;
-   bitbucket mirror is no longer maintained.

## 0.7.1 (2014-06-05)

-   Switch to setuptools;
-   upload wheel tp pypi;
-   check Python 3.4 compatibility.

## 0.7 (2013-10-13)

IntDAWG and IntCompletionDAWG are implemented.

## 0.6 (2013-03-23)

Use less shared state internally. This should fix thread-safety bugs and
make iterkeys/iteritems reenterant.

## 0.5.1 (2013-03-01)

Internal tweaks: memory usage is reduced; something is a bit faster,
something is a bit slower.

## 0.5 (2012-10-08)

Storage scheme is updated to match DAWG==0.5. This enables the
alphabetical ordering of `BytesDAWG` and `RecordDAWG` items.

In order to read `BytesDAWG` or `RecordDAWG` created with versions of
DAWG \< 0.5 use `payload_separator` constructor argument:

    >>> BytesDAWG(payload_separator=b'\xff').load('old.dawg')

## 0.3.1 (2012-10-01)

Bug with empty DAWGs is fixed.

## 0.3 (2012-09-26)

-   `iterkeys` and `iteritems` methods.

## 0.2 (2012-09-24)

`prefixes` support.

## 0.1 (2012-09-20)

Initial release.


tests/utils.py
import os
import zipfile
from pathlib import Path

DEV_DATA_PATH = Path(__file__).parent.parent / "dev_data"


def data_path(*args):
    """
    Returns a path to dev data
    """
    return os.path.join(DEV_DATA_PATH, *args)


def words100k():
    zip_name = data_path('words100k.txt.zip')
    zf = zipfile.ZipFile(zip_name)
    txt = zf.open(zf.namelist()[0]).read().decode('utf8')
    return txt.splitlines()


tests/test_dawg.py
import pickle
import tempfile

import pytest

import dawg_python
from .utils import data_path


def test_c_dawg_contains():
    dawg = pytest.importorskip("dawg")  # import dawg
    bin_dawg = dawg.IntDAWG({'foo': 1, 'bar': 2, 'foobar': 3})

    d = dawg_python.Dictionary()

    fd, path = tempfile.mkstemp()
    bin_dawg.save(path)

    with open(path, 'rb') as f:
        d.read(f)

    assert d.contains(b'foo')
    assert not d.contains(b'x')
    assert d.contains(b'foobar')
    assert d.contains(b'bar')


class TestCompletionDAWG:
    keys = ['f', 'bar', 'foo', 'foobar']

    def dawg(self):
        return dawg_python.CompletionDAWG().load(data_path('small', 'completion.dawg'))

    def test_contains(self):
        d = self.dawg()
        for key in self.keys:
            assert key in d

    def test_contains_bytes(self):
        d = self.dawg()
        for key in self.keys:
            assert key.encode('utf8') in d

    def test_keys(self):
        d = self.dawg()
        assert d.keys() == sorted(self.keys)

    def test_iterkeys(self):
        d = self.dawg()
        assert list(d.iterkeys()) == d.keys()

    def test_completion(self):
        d = self.dawg()

        assert d.keys('z') == []
        assert d.keys('b') == ['bar']
        assert d.keys('foo') == ['foo', 'foobar']

    def test_no_segfaults_on_invalid_file(self):
        d = self.dawg()
        fd, path = tempfile.mkstemp()
        with open(path, 'w') as f:
            f.write('foo')

        with pytest.raises(Exception):
            d.load(path)

    def test_empty_dawg(self):
        d = dawg_python.CompletionDAWG().load(data_path('small', 'completion-empty.dawg'))
        assert d.keys() == []

    def test_prefixes(self):
        d = self.dawg()
        assert d.prefixes("foobarz") == ["f", "foo", "foobar"]
        assert d.prefixes("x") == []
        assert d.prefixes("bar") == ["bar"]


class TestIntDAWG:
    payload = {'foo': 1, 'bar': 5, 'foobar': 3}

    def dawg(self):
        return dawg_python.IntDAWG().load(data_path('small', 'int_dawg.dawg'))

    def test_getitem(self):
        d = self.dawg()
        for key in self.payload:
            assert d[key] == self.payload[key]

        with pytest.raises(KeyError):
            d['fo']

    def test_pickling(self):
        d = self.dawg()

        data = pickle.dumps(d)
        d2 = pickle.loads(data)

        for key, value in self.payload.items():
            assert key in d2
            assert d[key] == value


class TestIntCompletionDawg(TestIntDAWG):
    def dawg(self):
        return dawg_python.IntCompletionDAWG().load(data_path('small', 'int_completion_dawg.dawg'))

    def test_completion_keys(self):
        assert self.dawg().keys() == sorted(self.payload.keys())

    def test_completion_keys_with_prefix(self):
        assert self.dawg().keys('fo') == ['foo', 'foobar']
        assert self.dawg().keys('foo') == ['foo', 'foobar']
        assert self.dawg().keys('foob') == ['foobar']
        assert self.dawg().keys('z') == []
        assert self.dawg().keys('b') == ['bar']

    def test_completion_items(self):
        assert self.dawg().items() == sorted(self.payload.items(), key=lambda r: r[0])


tests/test_fuzzy.py
import dawg_python

from .utils import words100k, data_path

words = words100k()
dawg = dawg_python.Dictionary.load(data_path('large', 'int_dawg.dawg'))


class TestDictionary:
    def test_contains(self):
        for word in words:
            assert dawg.contains(word.encode('utf8'))

    def test_find(self):
        for word in words:
            assert dawg.find(word.encode('utf8')) == len(word)


_prepare_dev_data.py
#!/usr/bin/env python
"""
Script for building test DAWGs.
"""
import struct

import dawg

from bench.utils import words100k
from tests.test_prediction import TestPrediction


def create_dawg():
    words = words100k()
    return dawg.DAWG(words)


def create_bytes_dawg():
    words = words100k()
    values = [struct.pack('<H', len(word)) for word in words]
    return dawg.BytesDAWG(zip(words, values))


def create_record_dawg():
    words = words100k()
    values = [ [len(word)] for word in words]
    return dawg.RecordDAWG('<H', zip(words, values))


def create_int_dawg():
    words = words100k()
    values = [len(word) for word in words]
    return dawg.IntDAWG(zip(words, values))


def create_int_completion_dawg():
    words = words100k()
    values = [len(word) for word in words]
    return dawg.IntCompletionDAWG(zip(words, values))


def build_test_data():

    dawg.CompletionDAWG(['f', 'bar', 'foo', 'foobar']).save('dev_data/small/completion.dawg')
    dawg.CompletionDAWG([]).save('dev_data/small/completion-empty.dawg')

    bytes_data =  (
        ('foo', b'data1'),
        ('bar', b'data2'),
        ('foo', b'data3'),
        ('foobar', b'data4')
    )
    dawg.BytesDAWG(bytes_data).save('dev_data/small/bytes.dawg')

    record_data = (
        ('foo',     (3, 2, 256)),
        ('bar',     (3, 1, 0)),
        ('foo',     (3, 2, 1)),
        ('foobar',  (6, 3, 0))
    )
    dawg.RecordDAWG(">3H", record_data).save('dev_data/small/record.dawg')

    int_data = {'foo': 1, 'bar': 5, 'foobar': 3}
    dawg.IntDAWG(int_data).save('dev_data/small/int_dawg.dawg')
    dawg.IntCompletionDAWG(int_data).save('dev_data/small/int_completion_dawg.dawg')

    dawg.DAWG(TestPrediction.DATA).save('dev_data/small/prediction.dawg')
    dawg.RecordDAWG("=H", [(k, (len(k),)) for k in TestPrediction.DATA]).save('dev_data/small/prediction-record.dawg')

    create_dawg().save('dev_data/large/dawg.dawg')
    create_bytes_dawg().save('dev_data/large/bytes_dawg.dawg')
    create_record_dawg().save('dev_data/large/record_dawg.dawg')
    create_int_dawg().save('dev_data/large/int_dawg.dawg')
    #create_int_completion_dawg().save('dev_data/large/int_completion_dawg.dawg')


if __name__ == '__main__':
    build_test_data()


dawg_python/dawgs.py
import struct
from binascii import a2b_base64

from . import wrapper


class DAWG:
    """
    Base DAWG wrapper.
    """

    def __init__(self):
        self.dct = None

    def __contains__(self, key):
        if not isinstance(key, bytes):
            key = key.encode('utf8')
        return self.dct.contains(key)

    def load(self, path):
        """
        Loads DAWG from a file.
        """
        self.dct = wrapper.Dictionary.load(path)
        return self

    def _has_value(self, index):
        return self.dct.has_value(index)

    def _similar_keys(self, current_prefix, key, index, replace_chars):

        res = []
        start_pos = len(current_prefix)
        end_pos = len(key)
        word_pos = start_pos

        while word_pos < end_pos:
            b_step = key[word_pos].encode('utf8')

            if b_step in replace_chars:
                next_index = index
                b_replace_char, u_replace_char = replace_chars[b_step]

                next_index = self.dct.follow_bytes(b_replace_char, next_index)

                if next_index is not None:
                    prefix = current_prefix + key[start_pos:word_pos] + u_replace_char
                    extra_keys = self._similar_keys(prefix, key, next_index, replace_chars)
                    res += extra_keys

            index = self.dct.follow_bytes(b_step, index)
            if index is None:
                break
            word_pos += 1

        else:
            if self._has_value(index):
                found_key = current_prefix + key[start_pos:]
                res.insert(0, found_key)

        return res

    def similar_keys(self, key, replaces):
        """
        Returns all variants of ``key`` in this DAWG according to
        ``replaces``.

        ``replaces`` is an object obtained from
        ``DAWG.compile_replaces(mapping)`` where mapping is a dict
        that maps single-char unicode sitrings to another single-char
        unicode strings.

        This may be useful e.g. for handling single-character umlauts.
        """
        return self._similar_keys("", key, self.dct.ROOT, replaces)

    @classmethod
    def compile_replaces(cls, replaces):

        for k, v in replaces.items():
            if len(k) != 1 or len(v) != 1:
                raise ValueError("Keys and values must be single-char unicode strings.")

        return dict(
            (
                k.encode('utf8'),
                (v.encode('utf8'), v),
            )
            for k, v in replaces.items()
        )

    def prefixes(self, key):
        """
        Returns a list with keys of this DAWG that are prefixes of the ``key``.
        """
        res = []
        index = self.dct.ROOT
        if not isinstance(key, bytes):
            key = key.encode('utf8')

        pos = 1

        for ch in key:
            index = self.dct.follow_char(ch, index)
            if not index:
                break

            if self._has_value(index):
                res.append(key[:pos].decode('utf8'))
            pos += 1

        return res


class CompletionDAWG(DAWG):
    """
    DAWG with key completion support.
    """

    def __init__(self):
        super().__init__()
        self.guide = None

    def keys(self, prefix=""):
        b_prefix = prefix.encode('utf8')
        res = []

        index = self.dct.follow_bytes(b_prefix, self.dct.ROOT)
        if index is None:
            return res

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, b_prefix)

        while completer.next():
            key = completer.key.decode('utf8')
            res.append(key)

        return res

    def iterkeys(self, prefix=""):
        b_prefix = prefix.encode('utf8')
        index = self.dct.follow_bytes(b_prefix, self.dct.ROOT)
        if index is None:
            return

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, b_prefix)

        while completer.next():
            yield completer.key.decode('utf8')

    def load(self, path):
        """
        Loads DAWG from a file.
        """
        self.dct = wrapper.Dictionary()
        self.guide = wrapper.Guide()

        with open(path, 'rb') as f:
            self.dct.read(f)
            self.guide.read(f)

        return self


PAYLOAD_SEPARATOR = b'\x01'
MAX_VALUE_SIZE = 32768


class BytesDAWG(CompletionDAWG):
    """
    DAWG that is able to transparently store extra binary payload in keys;
    there may be several payloads for the same key.

    In other words, this class implements read-only DAWG-based
    {unicode -> list of bytes objects} mapping.
    """

    def __init__(self, payload_separator=PAYLOAD_SEPARATOR):
        super().__init__()
        self._payload_separator = payload_separator

    def __contains__(self, key):
        if not isinstance(key, bytes):
            key = key.encode('utf8')
        return bool(self._follow_key(key))

    # def b_has_key(self, key):
    #     return bool(self._follow_key(key))

    def __getitem__(self, key):
        res = self.get(key)
        if res is None:
            raise KeyError(key)
        return res

    def get(self, key, default=None):
        """
        Returns a list of payloads (as byte objects) for a given key
        or ``default`` if the key is not found.
        """
        if not isinstance(key, bytes):
            key = key.encode('utf8')

        return self.b_get_value(key) or default

    def _follow_key(self, b_key):
        index = self.dct.follow_bytes(b_key, self.dct.ROOT)
        if not index:
            return False

        index = self.dct.follow_bytes(self._payload_separator, index)
        if not index:
            return False

        return index

    def _value_for_index(self, index):
        res = []

        completer = wrapper.Completer(self.dct, self.guide)

        completer.start(index)
        while completer.next():
            b64_data = completer.key
            res.append(a2b_base64(b64_data))

        return res

    def b_get_value(self, b_key):
        index = self._follow_key(b_key)
        if not index:
            return []
        return self._value_for_index(index)

    def keys(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')
        res = []

        index = self.dct.ROOT

        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return res

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            payload_idx = completer.key.index(self._payload_separator)
            u_key = completer.key[:payload_idx].decode('utf8')
            res.append(u_key)
        return res

    def iterkeys(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')

        index = self.dct.ROOT

        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            payload_idx = completer.key.index(self._payload_separator)
            u_key = completer.key[:payload_idx].decode('utf8')
            yield u_key

    def items(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')
        res = []

        index = self.dct.ROOT
        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return res

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            key, value = completer.key.split(self._payload_separator)
            res.append((key.decode('utf8'), a2b_base64(value)))

        return res

    def iteritems(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')

        index = self.dct.ROOT
        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            key, value = completer.key.split(self._payload_separator)
            item = (key.decode('utf8'), a2b_base64(value))
            yield item

    def _has_value(self, index):
        return self.dct.follow_bytes(PAYLOAD_SEPARATOR, index)

    def _similar_items(self, current_prefix, key, index, replace_chars):

        res = []
        start_pos = len(current_prefix)
        end_pos = len(key)
        word_pos = start_pos

        while word_pos < end_pos:
            b_step = key[word_pos].encode('utf8')

            if b_step in replace_chars:
                next_index = index
                b_replace_char, u_replace_char = replace_chars[b_step]

                next_index = self.dct.follow_bytes(b_replace_char, next_index)
                if next_index:
                    prefix = current_prefix + key[start_pos:word_pos] + u_replace_char
                    extra_items = self._similar_items(prefix, key, next_index, replace_chars)
                    res += extra_items

            index = self.dct.follow_bytes(b_step, index)
            if not index:
                break
            word_pos += 1

        else:
            index = self.dct.follow_bytes(self._payload_separator, index)
            if index:
                found_key = current_prefix + key[start_pos:]
                value = self._value_for_index(index)
                res.insert(0, (found_key, value))

        return res

    def similar_items(self, key, replaces):
        """
        Returns a list of (key, value) tuples for all variants of ``key``
        in this DAWG according to ``replaces``.

        ``replaces`` is an object obtained from
        ``DAWG.compile_replaces(mapping)`` where mapping is a dict
        that maps single-char unicode sitrings to another single-char
        unicode strings.
        """
        return self._similar_items("", key, self.dct.ROOT, replaces)

    def _similar_item_values(self, start_pos, key, index, replace_chars):
        res = []
        end_pos = len(key)
        word_pos = start_pos

        while word_pos < end_pos:
            b_step = key[word_pos].encode('utf8')

            if b_step in replace_chars:
                next_index = index
                b_replace_char, u_replace_char = replace_chars[b_step]

                next_index = self.dct.follow_bytes(b_replace_char, next_index)
                if next_index:
                    extra_items = self._similar_item_values(word_pos + 1, key, next_index, replace_chars)
                    res += extra_items

            index = self.dct.follow_bytes(b_step, index)
            if not index:
                break
            word_pos += 1

        else:
            index = self.dct.follow_bytes(self._payload_separator, index)
            if index:
                value = self._value_for_index(index)
                res.insert(0, value)

        return res

    def similar_item_values(self, key, replaces):
        """
        Returns a list of values for all variants of the ``key``
        in this DAWG according to ``replaces``.

        ``replaces`` is an object obtained from
        ``DAWG.compile_replaces(mapping)`` where mapping is a dict
        that maps single-char unicode sitrings to another single-char
        unicode strings.
        """
        return self._similar_item_values(0, key, self.dct.ROOT, replaces)


class RecordDAWG(BytesDAWG):
    def __init__(self, fmt, payload_separator=PAYLOAD_SEPARATOR):
        super().__init__(payload_separator)
        self._struct = struct.Struct(str(fmt))
        self.fmt = fmt

    def _value_for_index(self, index):
        value = super()._value_for_index(index)
        return [self._struct.unpack(val) for val in value]

    def items(self, prefix=""):
        res = super().items(prefix)
        return [(key, self._struct.unpack(val)) for (key, val) in res]

    def iteritems(self, prefix=""):
        res = super().iteritems(prefix)
        return ((key, self._struct.unpack(val)) for (key, val) in res)


LOOKUP_ERROR = -1


class IntDAWG(DAWG):
    """
    Dict-like class based on DAWG.
    It can store integer values for unicode keys.
    """

    def __getitem__(self, key):
        res = self.get(key, LOOKUP_ERROR)
        if res == LOOKUP_ERROR:
            raise KeyError(key)
        return res

    def get(self, key, default=None):
        """
        Return value for the given key or ``default`` if the key is not found.
        """
        if not isinstance(key, bytes):
            key = key.encode('utf8')
        res = self.b_get_value(key)
        if res == LOOKUP_ERROR:
            return default
        return res

    def b_get_value(self, key):
        return self.dct.find(key)


class IntCompletionDAWG(CompletionDAWG, IntDAWG):
    """
    Dict-like class based on DAWG.
    It can store integer values for unicode keys and support key completion.
    """

    def items(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')
        res = []
        index = self.dct.ROOT

        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return res

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            res.append((completer.key.decode('utf8'), completer.value()))

        return res

    def iteritems(self, prefix=""):
        if not isinstance(prefix, bytes):
            prefix = prefix.encode('utf8')
        index = self.dct.ROOT

        if prefix:
            index = self.dct.follow_bytes(prefix, index)
            if not index:
                return

        completer = wrapper.Completer(self.dct, self.guide)
        completer.start(index, prefix)

        while completer.next():
            yield completer.key.decode('utf8'), completer.value()


bench/speed.py
import random
import string
import timeit

import dawg_python
from .utils import data_path, words100k


def random_words(num):
    russian = 'абвгдеёжзиклмнопрстуфхцчъыьэюя'
    alphabet = f'{russian}{string.ascii_letters}'
    return [
        "".join([random.choice(alphabet) for x in range(random.randint(1, 15))])
        for _ in range(num)
    ]


def truncated_words(words):
    return [word[:3] for word in words]


def prefixes1k(words, prefix_len):
    words = [w for w in words if len(w) >= prefix_len]
    every_nth = int(len(words) / 1000)
    _words = [w[:prefix_len] for w in words[::every_nth]]
    return _words[:1000]


WORDS100k = words100k()
MIXED_WORDS100k = truncated_words(WORDS100k)
NON_WORDS100k = random_words(100000)
PREFIXES_3_1k = prefixes1k(WORDS100k, 3)
PREFIXES_5_1k = prefixes1k(WORDS100k, 5)
PREFIXES_8_1k = prefixes1k(WORDS100k, 8)
PREFIXES_15_1k = prefixes1k(WORDS100k, 15)


def format_result(key, value):
    print("%55s:    %s" % (key, value))


def bench(name, timer, descr='M ops/sec', op_count=0.1, repeats=3, runs=5):
    try:
        times = []
        for x in range(runs):
            times.append(timer.timeit(repeats))

        def op_time(time):
            return op_count * repeats / time

        val = f"{op_time(min(times)):0.3f}{descr}"
        format_result(name, val)
    except (AttributeError, TypeError):
        format_result(name, "not supported")


def load_dawg():
    return dawg_python.DAWG().load(data_path('large', 'dawg.dawg'))


def load_bytes_dawg():
    return dawg_python.BytesDAWG().load(data_path('large', 'bytes_dawg.dawg'))


def load_record_dawg():
    return dawg_python.RecordDAWG('<H').load(data_path('large', 'record_dawg.dawg'))


def load_int_dawg():
    return dawg_python.IntDAWG().load(data_path('large', 'int_dawg.dawg'))


def benchmark():
    print('\n====== Benchmarks (100k unique unicode words) =======\n')

    tests = [
        ('__getitem__ (hits)', "for word in WORDS100k: data[word]", 'M ops/sec', 0.1, 3),
        ('get() (hits)', "for word in WORDS100k: data.get(word)", 'M ops/sec', 0.1, 3),
        ('get() (misses)', "for word in NON_WORDS_10k: data.get(word)", 'M ops/sec', 0.01, 5),
        ('__contains__ (hits)', "for word in WORDS100k: word in data", 'M ops/sec', 0.1, 3),
        ('__contains__ (misses)', "for word in NON_WORDS100k: word in data", 'M ops/sec', 0.1, 3),
        ('items()', 'list(data.items())', ' ops/sec', 1, 1),
        ('keys()', 'list(data.keys())', ' ops/sec', 1, 1),
        # ('values()', 'list(data.values())', ' ops/sec', 1, 1),
    ]

    common_setup = """
from __main__ import load_dawg, load_bytes_dawg, load_record_dawg, load_int_dawg
from __main__ import WORDS100k, NON_WORDS100k, MIXED_WORDS100k
from __main__ import PREFIXES_3_1k, PREFIXES_5_1k, PREFIXES_8_1k, PREFIXES_15_1k
NON_WORDS_10k = NON_WORDS100k[:10000]
NON_WORDS_1k = ['ыва', 'xyz', 'соы', 'Axx', 'avы']*200
"""
    dict_setup = common_setup + 'data = dict((word, len(word)) for word in WORDS100k);'
    dawg_setup = common_setup + 'data = load_dawg();'
    bytes_dawg_setup = common_setup + 'data = load_bytes_dawg();'
    record_dawg_setup = common_setup + 'data = load_record_dawg();'
    int_dawg_setup = common_setup + 'data = load_int_dawg();'

    structures = [
        ('dict', dict_setup),
        ('DAWG', dawg_setup),
        ('BytesDAWG', bytes_dawg_setup),
        ('RecordDAWG', record_dawg_setup),
        ('IntDAWG', int_dawg_setup),
    ]
    for test_name, test, descr, op_count, repeats in tests:
        for name, setup in structures:
            timer = timeit.Timer(test, setup)
            full_test_name = f"{name} {test_name}"
            bench(full_test_name, timer, descr, op_count, repeats)

    # DAWG-specific benchmarks
    for struct_name, setup in structures[1:]:
        _bench_data = [
            ('hits', 'WORDS100k'),
            ('mixed', 'MIXED_WORDS100k'),
            ('misses', 'NON_WORDS100k'),
        ]

        for meth in ['prefixes']:
            for name, data in _bench_data:
                bench(
                    f'{struct_name}.{meth} ({name})',
                    timeit.Timer(
                        f"for word in {data}:\n"
                        f"   data.{meth}(word)",
                        setup
                    ),
                    runs=3
                )

        _bench_data = [
            ('xxx', 'avg_len(res)==415', 'PREFIXES_3_1k'),
            ('xxxxx', 'avg_len(res)==17', 'PREFIXES_5_1k'),
            ('xxxxxxxx', 'avg_len(res)==3', 'PREFIXES_8_1k'),
            ('xxxxx..xx', 'avg_len(res)==1.4', 'PREFIXES_15_1k'),
            ('xxx', 'NON_EXISTING', 'NON_WORDS_1k'),
        ]
        for xxx, avg, data in _bench_data:
            for meth in ['keys', 'items']:
                bench(
                    f'{struct_name}.{meth}(prefix="{xxx}"), {avg}',
                    timeit.Timer(
                        f"for word in {data}: data.{meth}(word)",
                        setup
                    ),
                    'K ops/sec',
                    op_count=1,
                    runs=3
                )

            for meth in ['iterkeys', 'iteritems']:
                bench(
                    f'{struct_name}.{meth}(prefix="{xxx}"), {avg}',
                    timeit.Timer(
                        f"for word in {data}: list(data.{meth}(word))",
                        setup
                    ),
                    'K ops/sec',
                    op_count=1,
                    runs=3
                )


if __name__ == '__main__':
    benchmark()
    # profiling()
    print('\n~~~~~~~~~~~~~~\n')
